Update uDiscovery specs (#118)
eclipse-uprotocol/up-spec: up-l3/udiscovery/v3/README.adoc

Reviewer: the other new APIs in the proto are not covered here and there is no explanation about their purpose.

== Purpose

> uE's (Software Entities) will be deployed dynamically on devices that are scattered throughout the uProtocol network. In order for these  uE's to know about what is available on their own device or on other devices, we need the ability to discover uEs, their locations and what they serve (resources, methods, messages, etc...). There are number of Internet protocols such as DNS, Zeroconf, and others, to discover devices and services running on said devices. Intent of uDiscovery is to reuse as much as possible, existing Internet standards for discovering uThings as well.

Reviewer: ```suggestion
uE's (Software Entities) will be deployed dynamically on devices that are scattered throughout the uProtocol network. In order for these uE's to know about what is available on their own device or on other devices, we need the ability to discover uEs, their locations and what they serve (resources, methods, messages, etc...). There are number of Internet protocols such as DNS, Zeroconf, and others, to discover devices and services running on said devices. Intent of uDiscovery is to reuse as much as possible, existing Internet standards for discovering uThings as well.
```

|Term |Definition
|DNS |Domain Name System covered in various IETF RFCs

> |CDS |Central Discovery Service implementation of uDiscovery

Reviewer: ```suggestion
|uCDS |Central Discovery Service implementation of uDiscovery
```

> |DDS | Domain Discovery Service(local to a domain) implemenation of uDiscovery

Reviewer: ```suggestion
|uDDS | Domain Discovery Service(local to a domain) implementation of uDiscovery
```

> |LDS |Local (to a device) implementation of uDiscovery

Reviewer: ```suggestion
|uLDS |Local (to a device) implementation of uDiscovery
```

> |Local Entity |Entity  who is local to the device (local information)

Reviewer: Remove the remote and local entity rows in the table

|Component |Description |Requirements

> |*Local uDiscovery Service (LDS)*

Reviewer: ```suggestion
|*Local uDiscovery Service (uLDS)*
```

|Service to provide discovery of uThings per device. Every LDS stores information of all local entities. In addition, it caches information for uEs on other devices.
a|* Every uDevice *MUST* have LDS

| *Domain uDiscovery Service (uDDS)*
| Service to provide discovery of all uThings per domain.
a|
* uDDS *MUST* store `uEs` for all the devices under a given domain hierarchy
* *MUST* be only one uDDS within a domain
* A uLDS *MAY* also function as uDDS and/or a uCDS for a uDomain or network
* uDDS *MUST* support queries from uLDSs that are within its domain
* uDDS *MUST* relay queries from uLDS to the uCDS and then cache the results per the TTL requirements from query API response
* uDDS *SHOULD* be located in the always-on uDevices

|*Central uDiscovery Service (uCDS)*
|Central repository storing all deployed uDevice information in the network of connected devices residing in the Cloud. It is a superset of the local uDiscovery database content
a|
* *MUST* have one CDS that is globally addressable by the various domain/device LDS
|===

> === Hierarchical Specifications

Reviewer: This section is no longer relavent and should be removed IMHO

> === Database Access

Reviewer: This should not be in the specification as it is an implementation detail that is not part of the API specifications (more on implementation detail for a given DS. Please remove, same with policy. I don't think we should put database specific details here. It could be a best practice section somewhere else.

In the following section we will explain the various APIs and interfaces that are defined in uDiscovery and their requirements. Interface definitions (input and output parameters, etc...) are covered in the link:../../../up-core-api/uprotocol/core/udiscovery/v3/udiscovery.proto[udiscovery.proto].

> === Notifications

Reviewer: No more notifications? If uEs are added or removed, how are we notified?

Query APIs are used to lookup content in the database, either to resolve URIs (to be used by applications) or to fetch content of a database.

* Any uE *MAY* call the query APIs defined in the sections below

> * Remote Nodes that are `expired` *MUST* be refreshed to the CDS

Reviewer: All node references need to be removed

* Locally `expired` Nodes *MUST NOT* be returned in a query

==== Find UE

Figure below illustrates the flows for performing a query to the LDS. Uri details for requested uE are returned if found in uLDS first, else the query is sent to uDDS and in turn to uCDS if required.
* *MUST* update `last_accessed` Node attribute when API is called

image:FindUE.svg[FinduE]

> ==== Get UE Topics

Reviewer: Please add more details as to why this APi exists, who would use it when

Get uE Topics is used to fetch the list of topics published by a uE. 

> * *MUST* update `last_accessed` Node attribute when API is called

Reviewer: No longer relevant for the API.

== Failure Recovery

> In the event that the databases between the CDS and LDS becomes out of sync, the discovery service components (uLDS, uDDS, uCDS) *MAY* fetch the contents using read APIs.

Reviewer: Database specific detail that is out of scope for this API documentation.
